File: 03-sliding-window/03-substring-with-concatenation-of-all-words/Solution.py
class Solution:
    def findSubstring(self, s: str, words: list[str]) -> list[int]:
        if not words or not s:
            return []
        word_len = len(words[0])
        num_words = len(words)
        total_len = word_len * num_words
        n = len(s)
        if n < total_len:
            return []

        word_count: dict[str, int] = {}
        for w in words:
            word_count[w] = word_count.get(w, 0) + 1

        res = []
        for offset in range(word_len):
            left = offset
            count = 0
            window: dict[str, int] = {}
            for right in range(offset, n - word_len + 1, word_len):
                word = s[right:right + word_len]
                if word in word_count:
                    window[word] = window.get(word, 0) + 1
                    count += 1
                    while window[word] > word_count[word]:
                        left_word = s[left:left + word_len]
                        window[left_word] -= 1
                        left += word_len
                        count -= 1
                    if count == num_words:
                        res.append(left)
                        left_word = s[left:left + word_len]
                        window[left_word] -= 1
                        left += word_len
                        count -= 1
                else:
                    window.clear()
                    count = 0
                    left = right + word_len
        return res

File: 03-sliding-window/03-substring-with-concatenation-of-all-words/Solution.java
import java.util.*;

class Solution {
    public List<Integer> findSubstring(String s, String[] words) {
        List<Integer> res = new ArrayList<>();
        if (words.length == 0 || s.isEmpty()) return res;
        int wordLen = words[0].length();
        int numWords = words.length;
        int totalLen = wordLen * numWords;
        int n = s.length();
        if (n < totalLen) return res;

        Map<String,Integer> wordCount = new HashMap<>();
        for (String w : words) wordCount.merge(w, 1, Integer::sum);

        for (int offset = 0; offset < wordLen; offset++) {
            int left = offset, count = 0;
            Map<String,Integer> window = new HashMap<>();
            for (int right = offset; right + wordLen <= n; right += wordLen) {
                String word = s.substring(right, right + wordLen);
                if (wordCount.containsKey(word)) {
                    window.merge(word, 1, Integer::sum);
                    count++;
                    while (window.get(word) > wordCount.get(word)) {
                        String leftWord = s.substring(left, left + wordLen);
                        window.merge(leftWord, -1, Integer::sum);
                        left += wordLen;
                        count--;
                    }
                    if (count == numWords) {
                        res.add(left);
                        String leftWord = s.substring(left, left + wordLen);
                        window.merge(leftWord, -1, Integer::sum);
                        left += wordLen;
                        count--;
                    }
                } else {
                    window.clear();
                    count = 0;
                    left = right + wordLen;
                }
            }
        }
        return res;
    }
}
